plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb

```diff
@@ -0,0 +1,150 @@
+    let(:params) { { target_usernames: %w[lechuck elaine] } }
+
+    subject(:contract) { described_class.new(params) }
+
+    it { is_expected.to validate_presence_of :target_usernames }
+
+    context "when the target_usernames argument is a string" do
+      let(:params) { { target_usernames: "lechuck,elaine" } }
+
+      it "splits it into an array" do
+        contract.validate
+        expect(contract.target_usernames).to eq(%w[lechuck elaine])
+      end
+    end
+  end
+
+  describe ".call" do
+    subject(:result) { described_class.call(params) }
+
+    fab!(:current_user) { Fabricate(:user, username: "guybrush") }
+    fab!(:user_1) { Fabricate(:user, username: "lechuck") }
+    fab!(:user_2) { Fabricate(:user, username: "elaine") }
+
+    let(:guardian) { Guardian.new(current_user) }
+    let(:params) { { guardian: guardian, target_usernames: %w[lechuck elaine] } }
+
+    before { Group.refresh_automatic_groups! }
+
+    context "when all steps pass" do
+      it "sets the service result as successful" do
+        expect(result).to be_a_success
+      end
+
+      it "creates the channel" do
+        expect { result }.to change { Chat::Channel.count }
+        expect(result.channel.chatable).to have_attributes(
+          user_ids: match_array([current_user.id, user_1.id, user_2.id]),
+        )
+      end
+
+      it "makes all target users a member of the channel and only updates the current user to following" do
+        expect { result }.to change { Chat::UserChatChannelMembership.count }.by(3)
+        expect(result.channel.user_chat_channel_memberships.pluck(:user_id)).to match_array(
+          [current_user.id, user_1.id, user_2.id],
+        )
+        result.channel.user_chat_channel_memberships.each do |membership|
+          expect(membership).to have_attributes(
+            following: true,
+            muted: false,
+            desktop_notification_level: "always",
+            mobile_notification_level: "always",
+          )
+        end
+      end
+
+      it "publishes the new channel" do
+        messages =
+          MessageBus.track_publish(Chat::Publisher::NEW_CHANNEL_MESSAGE_BUS_CHANNEL) { result }
+        expect(messages.first.data[:channel][:title]).to eq("@elaine, @lechuck")
+      end
+
+      context "when there is an existing direct message channel for the target users" do
+        before { described_class.call(params) }
+
+        it "does not create a channel" do
+          expect { result }.to not_change { Chat::Channel.count }.and not_change {
+                  Chat::DirectMessage.count
+                }
+        end
+
+        it "does not double-insert the channel memberships" do
+          expect { result }.not_to change { Chat::UserChatChannelMembership.count }
+        end
+      end
+    end
+
+    context "when the current user cannot make direct messages" do
+      before { SiteSetting.direct_message_enabled_groups = Fabricate(:group).id }
+
+      fab!(:current_user) { Fabricate(:user) }
```

Comment:
`fab!` and `let` should appear before `before` blocks.

Reply:
I wish we could have rubocop rules for this and the subject one

Reply:
Going to check if they already exists :slightly_smiling_face:

Reply:
I haven’t found an existing rule for `let` and `before` blocks, but there is one for `subject`: https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#rspecleadingsubject
I’m going to open a PR against our rubocop gem and see if this is something we could merge quickly.

Reply:
I’ve opened two PRs:
https://github.com/discourse/rubocop-discourse/pull/32
https://github.com/discourse/discourse/pull/22225